apis/hive/v1/metricsconfig/metricsConfig.go

```diff
@@ -4,4 +4,10 @@ type MetricsConfig struct {
 	// Optional metrics and their configurations
 	// +optional
 	MetricsWithDuration []MetricsWithDuration `json:"metricsWithDuration"`
+	// MetricsWithClusterTypeLabels is a map of cluster type to the label applied on the cluster deployment.
+	// This map is meant to be used when it would be helpful to label certain metrics with more cluster type labels.
+	// If the provided label is present on a cluster deployment, all metrics that are logged per cluster type will
+	// attempt to label with the provided cluster type as well.
+	// +optional
+	MetricsWithClusterTypeLabels map[string]string `json:"metricsWithClusterTypeLabels"`
```

Comment:
_bikesheds_

The `MetricsWith` prefix indicates that the field will contain a list of metrics (in the case of `MetricsWithDuration`, they're identified by name), so probably not appropriate here.

`ClusterType` isn't as precise as I would like because a) we already have a metric using a label called `cluster-type`, b) "type" is like a super vague word anyway, and c) there's nothing restricting the user to just cluster-type-ish labels -- e.g. they could use `version-major-minor-patch` (I guess you could argue that version is a "type", but see b).

I'm open to other suggestions, but I'm thinking `AdditionalClusterDeploymentLabels`. This helps indicate that the labels are coming from ClusterDeployment objects; and that they'll be included in addition to whatever baseline labels we have hardcoded. (Even if we decide to fold `hive.openshift.io/cluster-type` into these, we still have the library-ish ones like `job` and `instance`, so "Additional" is still accurate.)

Also, as discussed in slack, this docstring ought to include the caveat that labels with unbounded values, such as cluster namespaces, names, IDs, etc., should be avoided due to cardinality concerns.

Also, we need _some_ way of documenting which metrics are subject to this feature. I can't think of a great way to do this that's guaranteed not to go stale... can you? Even pointing to the code is fraught -- there's _already_ two different places we set them up.

Also, should this be `omitempty`?

As for the docstring's explanation of what this field does, I'm trying to figure out how to wordsmith it to be a bit clearer. Here's an attempt:

AdditionalClusterDeploymentLabels allows configuration of additional labels to be applied to certain metrics. The keys can be any string value suitable for a metric label (see [link to prom docs describing that format]). The values can be any ClusterDeployment label key (from `metadata.labels`). When observing an affected metric, hive will label it with the specified metric key, and copy the value from the specified ClusterDeployment label. For example, including `{"ocp_major_version": "hive.openshift.io/version-major"}` will cause affected metrics to include a label key `ocp_major_version` with the value from the `hive.openshift.io/version-major` ClusterDeployment label -- e.g. "4".
NOTE: Avoid ClusterDeployment labels whose values are unbounded, such as those representing cluster names or IDs, as these will cause your prometheus database to grow indefinitely.
[And here's where we should list or point to which metrics are affected.]

Reply:
The point of `omitempty` is to avoid returning the field from etcd if it's in the schema (as understood by k8s) but not set in the database. This is to avoid compatibility problems where we retrieve such an object and try to unmarshal/use it in an older version of code where the (compiled-in) schema lacks that field. So the only way we get into trouble in our case is if we've installed the new CRD but are still running the old code -- we would panic as soon as we try to retrieve the HiveConfig object.

I'm going to say the lack of `omitempty` on `MetricsWithDuration` was an oversight rather than intentional, and we should have it on both fields.

Reply:
I missed this in one of the previous changes - added